pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,664 @@
+        - namewithpins  (an exact version pin is kept if any)
+        - dependencies  (a list of dependencies)
+    """
+    output_folder = tempfile.mkdtemp(prefix="pythonpackage-metafolder-")
+    try:
+        extract_metainfo_files_from_package(
+            dependency, output_folder, debug=debug
+        )
+
+        with open(os.path.join(output_folder, "METADATA"),
+                  "r", encoding="utf-8"
+                 ) as f:
+            # Get metadata and cut away description (is after 2 linebreaks)
+            metadata_entries = f.read().partition("\n\n")[0].splitlines()
+
+        if extract_type == "name" or extract_type == "namewithpins":
+            name = None
+            for meta_entry in metadata_entries:
+                if meta_entry.lower().startswith("name:"):
+                    return meta_entry.partition(":")[2].strip()
+            if name is None:
+                raise ValueError("failed to obtain package name")
+
+            if extract_type == "namewithpins" and \
+                    dependency.find("==") > 0 and \
+                    dependency.partition("==")[0].lower() == name.lower():
+                return name + "==" + dependency.partition("==")[2].lstrip()
+            return name
+        elif extract_type == "dependencies":
+            requirements = []
+            if os.path.exists(os.path.join(output_folder, 'pyproject.toml')):
+                with open(os.path.join(output_folder, 'pyproject.toml')) as f:
+                    build_sys = pytoml.load(f)['build-system']
+                    if "requires" in build_sys:
+                        requirements += build_sys["requires"]
+
+            # Add requirements from metadata:
+            requirements += [
+                entry.rpartition("Requires-Dist:")[2].strip()
+                for entry in metadata_entries
+                if entry.startswith("Requires-Dist")
+            ]
+
+            return list(set(requirements))  # remove duplicates
+    finally:
+        shutil.rmtree(output_folder)
+
+
+package_name_cache = dict()
+
+
+def get_package_name(dependency,
+                     keep_version_pins=False,
+                     use_cache=True):
+    def timestamp():
+        try:
+            return time.monotonic()
+        except AttributeError:
+            return time.time()  # Python 2.
+    try:
+        value = package_name_cache[dependency]
+        if value[0] + 600.0 > timestamp() and use_cache:
+            return value[1]
+    except KeyError:
+        pass
+    result = _extract_info_from_package(dependency, extract_type="name")
+    package_name_cache[dependency] = (timestamp(), result)
+    return result
+
+
+def get_package_dependencies(package, recursive=False, verbose=False):
+    """ Obtain the dependencies from a package. Please note this
+        function is possibly SLOW, especially if you enable
+        the recursive mode.
+    """
+    packages_processed = set()
+    package_queue = [package]
+    reqs = []
```

Comment:
why don't we use a set from the beginning then if at the end we want it to act as a set?

Reply:
oh yeah that was silly of me, changing it now :+1:

Reply:
changed :heavy_check_mark: